src/odb/src/db/dbScanChain.cpp

```diff
@@ -80,79 +85,222 @@ void _dbScanChain::differences(dbDiff& diff,
 
   DIFF_END
 }
 
 _dbScanChain::_dbScanChain(_dbDatabase* db)
 {
+  scan_partitions_ = new dbTable<_dbScanPartition>(
+      db,
+      this,
+      (GetObjTbl_t) &_dbScanChain::getObjectTable,
+      dbScanPartitionObj);
+  scan_insts_ = new dbTable<_dbScanInst>(
+      db, this, (GetObjTbl_t) &_dbScanChain::getObjectTable, dbScanInstObj);
 }
 
 _dbScanChain::_dbScanChain(_dbDatabase* db, const _dbScanChain& r)
 {
   name_ = r.name_;
-  length_ = r.length_;
   scan_in_ = r.scan_in_;
   scan_out_ = r.scan_out_;
-  scan_clock_ = r.scan_clock_;
   scan_enable_ = r.scan_enable_;
   test_mode_ = r.test_mode_;
+  scan_partitions_
+      = new dbTable<_dbScanPartition>(db, this, *r.scan_partitions_);
+  scan_insts_ = new dbTable<_dbScanInst>(db, this, *r.scan_insts_);
 }
 
 dbIStream& operator>>(dbIStream& stream, _dbScanChain& obj)
 {
   stream >> obj.name_;
-  stream >> obj.length_;
-  stream >> obj.cells_;
   stream >> obj.scan_in_;
   stream >> obj.scan_out_;
-  stream >> obj.scan_clock_;
   stream >> obj.scan_enable_;
   stream >> obj.test_mode_;
-  stream >> obj.partitions_;
-  stream >> obj.scan_insts_;
+  stream >> *obj.scan_partitions_;
+  stream >> *obj.scan_insts_;
   return stream;
 }
 
 dbOStream& operator<<(dbOStream& stream, const _dbScanChain& obj)
 {
   stream << obj.name_;
-  stream << obj.length_;
-  stream << obj.cells_;
   stream << obj.scan_in_;
   stream << obj.scan_out_;
-  stream << obj.scan_clock_;
   stream << obj.scan_enable_;
   stream << obj.test_mode_;
-  stream << obj.partitions_;
-  stream << obj.scan_insts_;
+  stream << *obj.scan_partitions_;
+  stream << *obj.scan_insts_;
   return stream;
 }
 
+dbObjectTable* _dbScanChain::getObjectTable(dbObjectType type)
+{
+  switch (type) {
+    case dbScanPartitionObj:
+      return scan_partitions_;
+    case dbScanInstObj:
+      return scan_insts_;
+    default:
+      break;
+  }
+  return getTable()->getObjectTable(type);
+}
+
+_dbScanChain::~_dbScanChain()
+{
+  delete scan_partitions_;
+  delete scan_insts_;
+}
+
 ////////////////////////////////////////////////////////////////////
 //
 // dbScanChain - Methods
 //
 ////////////////////////////////////////////////////////////////////
 
+dbSet<dbScanPartition> dbScanChain::getScanPartitions() const
+{
+  _dbScanChain* obj = (_dbScanChain*) this;
+  return dbSet<dbScanPartition>(obj, obj->scan_partitions_);
+}
+
+dbSet<dbScanInst> dbScanChain::getScanInsts() const
+{
+  _dbScanChain* obj = (_dbScanChain*) this;
+  return dbSet<dbScanInst>(obj, obj->scan_insts_);
+}
+
+// User Code Begin dbScanChainPublicMethods
+
+std::string_view dbScanChain::getName() const
+{
+  _dbScanChain* scan_chain = (_dbScanChain*) this;
+  return scan_chain->name_;
+}
+
+void dbScanChain::setName(std::string_view name)
+{
+  _dbScanChain* scan_chain = (_dbScanChain*) this;
+  scan_chain->name_ = name;
+}
+
+dbBTerm* _dbScanChain::getPin(_dbScanChain* scan_chain,
+                              dbId<dbBTerm> _dbScanChain::*field)
```

Comment:
Why not `const dbId<dbBTerm>& bterm_id`?  It would remove the need for the cast below.

Reply:
Makes sense, thanks!

I will be back in a few days as now I am realizing that I need to also support iterms as scan_in and scan_out for the scan chains for future scan compression implementation. I will tag you later once this is ready for another review, thanks!